compiler/next/include/chpl/Queries/Context.h

```diff
@@ -0,0 +1,344 @@
+
+  template<typename ResultType, typename... ArgTs>
+  chpl::querydetail::QueryMapResult<ResultType>*
+  updateResultForQuery(
+      const std::tuple<ArgTs...>& tupleOfArgs,
+      ResultType result,
+      bool& changedOut,
+      chpl::querydetail::QueryMap<ResultType,ArgTs...>* queryMap);
+
+  template<typename ResultType, typename... ArgTs>
+  chpl::querydetail::QueryMapResult<ResultType>*
+  updateResultForQuery(const std::tuple<ArgTs...>& tupleOfArgs,
+                       ResultType result,
+                       bool& changedOut,
+                       UniqueString queryName);
+
+  // Future Work: support marking used strings and garbage collecting the rest
+  // Could store an atomic uint_8 just after the string for the mark.
+
+  // Future Work: make the context thread-safe
+
+  // Future Work: allow moving some AST to a different context
+  //              (or, at least, that can handle the unique strings)
+
+  // Performance: Add fine-grained timing to measure
+  //  * the total time spent in each query
+  //  * the time spent in each query in Context functions
+  //    (i.e. hashtable manipulations)
+  //  * the time spent in each query in other queries
+  //  * the time spent in each query in other query code
+
+  // Performance: How can we arrange for better locality of reference/
+  // cache reuse for the maps from IDs? The IDs within a function could
+  // be just stored in a vector, but that would add an indirection to
+  // the hashtable. Is there a way to adjust the hashing function and
+  // tune the hashtable bucket size, or something? Do we need a custom
+  // hashtable?
+
+ public:
+  /**
+    Create a new AST Context.
+   */
+  static owned<Context> build();
+  ~Context();
+
+  /**
+    Get or create a unique string for a NULL-terminated C string
+    and return it as a C string. If the passed string is NULL,
+    this function will return an empty string.
+
+    Strings returned by this function will always be aligned to 2 bytes.
+
+    The function `UniqueString::build` returns such a string
+    with a wrapper type. It should be preferred for type safety
+    and to reduce redundant checks.
+   */
+  const char* uniqueCString(const char* s);
+
+  /**
+    Return the name of the module containing this ID.
+   */
+  UniqueString moduleNameForID(ID id);
+  /**
+    Return the file path for the file containing this ID.
+   */
+  UniqueString filePathForID(ID id);
+
+  /**
+    Query to get a file path given a module name
+   */
+  UniqueString filePathForModuleName(UniqueString modName);
+
+  /**
+    This function increments the current revision number stored
+    in the context. After it is called, the setters below can
+    be used to provide the input at that revision.
+
+    If the prepareToGC argument is true, when processing queries
+    in that revision, will prepare to garbage collect (by marking
+    elements appropriately).
+   */
+  void advanceToNextRevision(bool prepareToGC);
```

Comment:
Reading the PR message and the code thus far, I am trying to understand what a revision is in practice. Every time we compile (or IDE does on-the-fly compilation, for example) we bump the revision number, is that it? And our new AST nodes will know the last revision that edited them, in someway? (At least this is what I am assuming right now)

Reply:
Reading more, I think I was wrong. The revisions has nothing to do with ASTs but they are just there to facilitate the memoization of query results.

Reply:
The way I look at is that we update the revision number when the input has changed or might have changed. So for an IDE, that would be on a "code edit" event. In an incremental compilation framework, we might increment the revision number every time we try to compile something.